### DISCUSSION TITLE
dependsOn not working as expected for fields with the same attribute name

### DISCUSSION THREAD

**Discussion by ThaDaVos** [laravel/nova-issues#4953] [Q&A] [answered]
- Laravel Version: 9.35.1
- Nova Version: 4.16.1
- PHP Version: 8.0.23
- Database Driver & Version: MySQL 8.x
- Operating System and Version: Ubuntu
- Browser type and version: Firefox
- Reproduction Repository: https://github.com/###/###

### Description:
As discussed in https://github.com/laravel/nova-issues/issues/4648 there is an issue when using `dependsOn` with fields using the same attribute - in our case we're using two selects depending on a `type` field - based on the type a different select with different label is used - but due this issue, the first field is displayed twice in it's full.

### Detailed steps to reproduce the issue on a fresh Nova installation:
See: https://github.com/laravel/nova-issues/issues/4648

```PHP
Select::make('Field Type')
    ->options([
        'text' => 'Text',
        'number' => 'Number',
    ])
    ->onlyOnForms()
    ->displayUsingLabels(),

Number::make('Number', 'field_value')
    ->help('Number test')
    ->onlyOnForms()
    ->dependsOn(
        ['field_type'],
        function (Number $field, NovaRequest $request, FormData $formData) {
            if ($formData->field_type == 'number') {
                $field
                    ->show()
                    ->rules('required');
            } else {
                $field
                    ->hide();
            }
        }
    ),

Text::make('Text', 'field_value')
    ->help('Text test')
    ->onlyOnForms()
    ->dependsOn(
        ['field_type'],
        function (Text $field, NovaRequest $request, FormData $formData) {
            if ($formData->field_type == 'text') {
                $field
                    ->show()
                    ->rules('required');
            } else {
                $field
                    ->hide();
            }
        }
    ),
```

**Comment by crynobone** [comment]
Please add reproducing code.

**Comment by ThaDaVos** [comment]
See above - hope that's good

**Comment by crynobone** [comment]
https://user-images.githubusercontent.com/172966/196401580-9f8056b3-e843-463b-96c8-6920359e1dc8.mp4

Unable to reproduce the issue, please provide full reproducing repository based on fresh installation as suggested in the bug report template (or you can refer to https://github.com/nova-issues/issue-4936/commit/5b9b11918c9b8a50dfddea19fbaca427540fc7a1 for example)

**Comment by szymon-szychulski** [comment]
I have the same issue. Code to reproduce: 

```
Select::make('Action', 'action')
    ->options(
        array_combine(
            ['option_1', 'option_2'],
            ['Option 1', 'Option 2']
        )
    )
    ->displayUsingLabels(),


Select::make('Choose answer', 'answer')
    ->hide()
    ->options(array_combine(
        ['option_1', 'option_2'],
        ['Option 1', 'Option 2']
    ))
    ->dependsOn(
        ['action'],
        function (Select $field, NovaRequest $request, FormData $formData) {
            if ($formData->action === 'option_1') {
                $field->show();
            }
        }
    ),

Number::make('Enter number', 'answer')
    ->hide()
    ->dependsOn(
        ['action'],
        function (Select $field, NovaRequest $request, FormData $formData) {
            if ($formData->action === 'option_1') {
                $field->show();
            }
        }),
```

I hope it's clear now. In this case the attribute for second Select field and Number field is the same and Nova has the problem with displaying it correctly and handling. Of course there should be possibility to have two or more fields with the same name and using both of them with dependsOn.

**Comment by crynobone** [comment]
> Nova has the problem with displaying it correctly and handling.

Explain what you mean here.

**Comment by ThaDaVos** [comment]
> > Nova has the problem with displaying it correctly and handling.
> 
> Explain what you mean here.

Based on the issue he's commenting on - I bet the problem with displaying and handling @szymon-szychulski is talking about is the duplicate rendering of the fields - maybe it's an issue isolated to the `Select` field as your (@crynobone ) code doesn't have the issue

**Comment by ThaDaVos** [comment]
> I have the same issue. Code to reproduce:
> 
> ```
> Select::make('Action', 'action')
>     ->options(
>         array_combine(
>             ['option_1', 'option_2'],
>             ['Option 1', 'Option 2']
>         )
>     )
>     ->displayUsingLabels(),
> 
> 
> Select::make('Choose answer', 'answer')
>     ->hide()
>     ->options(array_combine(
>         ['option_1', 'option_2'],
>         ['Option 1', 'Option 2']
>     ))
>     ->dependsOn([
>         'action',
>         function (Select $field, NovaRequest $request, FormData $formData) {
>             if ($formData->action === 'option_1') {
>                $field->show();
>             }
>         }
>     ]),
> 
> Number::make('Enter number', 'answer')
>     ->hide()
>     ->dependsOn([
>         'action',
>         function (Number $field, NovaRequest $request, FormData $formData) {
>             if ($formData->action === 'option_2') {
>                $field->show();
>             }
>         }
>     ]),
> ```
> 
> I hope it's clear now. In this case the attribute for second Select field and Number field is the same and Nova has the problem with displaying it correctly and handling. Of course there should be possibility to have two or more fields with the same name and using both of them with dependsOn.

@szymon-szychulski - let me guess, it's the `Select` which renders twice?

**Comment by szymon-szychulski** [comment]
>Explain what you mean here.

Ok, I will show more complex example: 

```
    public function fields(Request $request): array
    {
        return [
            Select::make('Action', 'action')
                ->required()
                ->options(
                    array_combine(
                        AuditSpecificationCustomRuleModel::ACTIONS,
                        AuditSpecificationCustomRuleModel::ACTIONS
                    )
                )
                ->displayUsingLabels(),

            Select::make('Surcharge', 'action_parameters->0')
                ->hide()
                ->required()
                ->rules([
                    Rule::in(AuditSurchargeModel::SURCHARGES)
                ])
                ->options(array_combine(
                    AuditSurchargeModel::SURCHARGES,
                    AuditSurchargeModel::SURCHARGES
                ))
                ->dependsOn(
                    ['action'],
                    function (Select $field, NovaRequest $request, FormData $formData) {
                        if ($formData->action === AuditSpecificationCustomRuleModel::ADD_SURCHARGE) {
                            $field->show();
                        }
                    }
                ),

            Select::make('Price type', 'action_parameters->1')
                ->hide()
                ->required()
                ->rules([
                    Rule::in(MasterPriceListPrice::TYPES)
                ])
                ->options(array_combine(
                    MasterPriceListPrice::TYPES,
                    MasterPriceListPrice::TYPES
                ))
                ->dependsOn(
                    ['action'],
                    function (Select $field, NovaRequest $request, FormData $formData) {
                        if ($formData->action === AuditSpecificationCustomRuleModel::ADD_SURCHARGE) {
                            $field->show();
                        }
                    }
                ),

            Number::make('Price value max', 'action_parameters->2')
                ->hide()
                ->required()
                ->rules([
                    'numeric',
                    'min:0'
                ])
                ->dependsOn(
                    ['action'],
                    function (Number $field, NovaRequest $request, FormData $formData) {
                        if ($formData->action === AuditSpecificationCustomRuleModel::ADD_SURCHARGE) {
                            $field->show();
                        }
                    }
                ),

            Select::make('Shipment type', 'action_parameters->0')
                ->hide()
                ->required()
                ->rules([
                    Rule::in(OrderModel::TYPES)
                ])
                ->options(array_combine(
                    OrderModel::TYPES,
                    OrderModel::TYPES
                ))
                ->dependsOn(
                    ['action'],
                    function (Select $field, NovaRequest $request, FormData $formData) {
                        if ($formData->action === AuditSpecificationCustomRuleModel::SET_SHIPMENT_TYPE) {
                            $field->show();
                        }
                    }
                )
        ];
    }
```

Behaviour after I chose "Add surcharge" option: 

![image](https://user-images.githubusercontent.com/28821677/196666010-85af78b6-bf26-4fc9-a401-1586e7a14c8c.png)

Behaviour after I chose "Set shipment type" option:

![image](https://user-images.githubusercontent.com/28821677/196666204-5a0c47b9-f4e3-4b73-be6a-b0719012043d.png)

Basically fields dependent on action are displayed completely wrong. For "Add surcharge" option you can see both also last Select, where it should not be displayed, and for option "Set shipment type" no field is displayed.

**Comment by crynobone** [comment, marked as answer]
What's the return type of `AuditSpecificationCustomRuleModel::ADD_SURCHARGE` and wouldn't `Select::make('Surcharge', 'action_parameters->0')` and `Select::make('Shipment type', 'action_parameters->0')` set as single `Select` and you handle everything under `dependsOn`. That's how you do it. 

Having multiple `Select` with same attribute name will not be supported in Nova.

**Reply by ThaDaVos** [reply]
Why isn't having multiple `Select` with the same attribute through `dependsOn` not supported (as for example the `title` should be different and the logic inside `dependsOn` can stay single purpose), but is having the same attribute used, by a `Text` and `Number`  with `dependsOn`, supported?

Either, don't support it at all, or support all cases - at least, that's my opinion - seeing that multiple people are using the pattern of multiple `Select` fields, same attribute and `dependsOn` - seems that it's a case which gets used, in our case to keep our `dependsOn` concise, instead of using multiple if statements to further determine the state - also, it eases the other parts of the field's configuration, like the `title/label` but also `helpText` etc

**Reply by crynobone** [reply]
because it's impossible to target which field should be updated.
